test/opt/test_passes.cpp

```diff
@@ -0,0 +1,195 @@
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+#include <algorithm>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include <gtest/gtest.h>
+
+#include "opt/libspirv.hpp"
+#include "opt/pass_manager.h"
+#include "opt/passes.h"
+
+namespace {
+
+using namespace spvtools;
+
+// Concatenates a vector of strings into one string. Each string is postfixed
+// with the given |delimiter|.
+std::string Join(const std::vector<std::string>& insts, char delimiter = '\n') {
+  std::ostringstream oss;
+  for (const auto& inst : insts) oss << inst << delimiter;
+  return oss.str();
+}
+
+// Concatenates a vector of strings into one string. Each string is postfixed
+// with the given |delimiter|. If a string contains opcode for debug
+// instruction, that string will be ignored.
+std::string SkipDbgInstJoin(const std::vector<std::string>& insts,
+                            char delimiter = '\n') {
+  // Well, another place requiring the knowledge of the grammar and can be stale
+  // when SPIR-V is updated. It would be nice to automatically generate this,
+  // but the cost is just too high.
+  const std::vector<std::string> dbg_opcodes{
+      // clang-format off
+    "OpSourceContinued", "OpSource", "OpSourceExtension",
+    "OpName", "OpMemberName", "OpString",
+    "OpLine", "OpNoLine", "OpModuleProcessed"
+      // clang-format on
+  };
+
+  std::ostringstream oss;
+  for (const auto& inst : insts) {
+    if (std::all_of(dbg_opcodes.cbegin(), dbg_opcodes.cend(),
+                    [&inst](const std::string& op) {
+                      return inst.find(op) == std::string::npos;
+                    })) {
+      oss << inst << delimiter;
+    }
+  }
+  return oss.str();
+}
+
+// Template class for testing passes. It contains some handy utility methods for
+// running passes and checking results.
+//
+// To write value-Parameterized tests:
+//   using ValueParamTest = PassTest<::testing::TestWithParam<std::string>>;
+// To use as normal fixture:
+//   using FixtureTest = PassTest<::testing::Test>;
+template <typename GT>
+class PassTest : public GT {
+ public:
+  PassTest() : tools_(SPV_ENV_UNIVERSAL_1_1) {}
+
+  // Adds a pass to be run.
+  template <typename PassT>
+  void AddPass() {
+    manager_.AddPass<PassT>();
+  }
+
+  // Runs the pass on the binary assembled from the |original| assembly, and
+  // checks whether the optimized binary can be disassembled to the |expected|
+  // assembly.
```

Comment:
I suggest that the blurb should suggest use of SCOPED_TRACE so that when a failure occurs inside this method, then the developer can find out which caller was responsible.